plugins/ccgx/fu-ccgx-cyacd-file.c

```diff
@@ -0,0 +1,456 @@
+					current_handle->buffer = &buffer [buffer_pos];
+					next_handle_index++;
+				} else {
+					break;
+				}
+			}
+		}
+		line_pos++;
+		buffer_pos++;
+
+		if (current_handle != NULL) {
+			current_handle->buffer_size++;
+		}
+
+		if (ch == '\n' || ch == '\r') {
+			is_cr_lf = TRUE;
+		} else {
+			is_cr_lf = FALSE;
+		}
+	}
+	return next_handle_index;
+}
+
+/**
+ * fu_ccgx_cyacd_file_set_pos
+ *
+ * @handle: Cyacd File Handle
+ * @pos: cyacd data buffer
+ *
+ * Set position in cyacd buffer in cyacd handle
+ */
+void
+fu_ccgx_cyacd_file_set_pos (CyacdFileHandle *handle, guint32 pos)
+{
+	if (handle != NULL)
+		handle->pos = pos;
+}
+
+/**
+ * fu_ccgx_cyacd_file_get_pos:
+ * @handle: Cyacd File Handle
+ *
+ * Get position in cyacd buffer in cyacd handle
+ *
+ * Returns: position in n cyacd buffer
+ */
+
+guint32
+fu_ccgx_cyacd_file_get_pos (CyacdFileHandle *handle)
+{
+	if (handle == NULL)
+		return 0;
+	return handle->pos;
+}
+
+static gboolean
+cyacd_buffer_read_header (CyacdFileHandle *handle, guint16 *silicon_id)
+{
+	guint8 hex_buffer[4] = {0};
+	g_autofree guint8 *ascii_buffer = g_malloc0 (CYACD_ROW_ASCII_BUFFER_SIZE);
+
+	g_return_val_if_fail (ascii_buffer != NULL , FALSE);
+	g_return_val_if_fail (silicon_id != NULL , FALSE);
+	fu_ccgx_cyacd_file_set_pos (handle, 0);
+
+	if (cyacd_buffer_read_line (handle, ascii_buffer, CYACD_ROW_ASCII_BUFFER_SIZE)) {
+		convert_ascii_to_hex (ascii_buffer, sizeof(ascii_buffer), hex_buffer, sizeof(hex_buffer));
+		*silicon_id = hex_buffer[0]<<8 | hex_buffer[1];
+		return TRUE;
+	}
+	return FALSE;
+}
+
+static guint8
+calculate_byte_checksum(guint8 *ptr, guint32 size)
+{
+	guint8 checksum = 0;
+	guint32 index;
+
+	g_return_val_if_fail (ptr != NULL && size !=0, 0);
+
+	/* calculate the binary sum of all the data */
+	for (index = 0; index < size; index++)
+		checksum += ptr[index];
+
+	/* return the 2's complement of the binary sum */
+	return ((guint8)(1u) + (guint8)(~checksum));
+}
+
+static gboolean
+cyacd_buffer_read_row_raw_data(CyacdFileHandle *handle, guint8 *data, guint32 size)
+{
+	g_autofree guint8 *ascii_buffer = g_malloc0(CYACD_ROW_ASCII_BUFFER_SIZE);
+	guint32 read_size = 0;
+
+	g_return_val_if_fail (ascii_buffer != NULL, 0);
+	read_size = cyacd_buffer_read_line(handle, ascii_buffer, CYACD_ROW_ASCII_BUFFER_SIZE);
+
+	if (read_size > 0) {
+		if ((ascii_buffer[0] == ':') && (ascii_buffer[1] == '0') && (ascii_buffer[2] == '0')) {
+			if (size > ((read_size - 1) / 2))  /* 1 = 1 (:) */
+				size = (read_size - 1) / 2;
+
+			/* convert ASCII to HEX data and ignoring first entry */
+			convert_ascii_to_hex(&ascii_buffer[1], read_size - 1, data, size);
+			return TRUE;
+		}
+	}
+	return FALSE;
+}
+
+/**
+ * fu_ccgx_cyacd_file_parse:
+ * @handle: Cyacd File Handle
+ * @info[out] Cyacd file information
+ *
+ * Parse cyacd file and return cyacd file information
+ *
+ * Returns: %TRUE for success
+ */
+gboolean
+fu_ccgx_cyacd_file_parse (CyacdFileHandle *handle, CyacdFileInfo *info)
```

Comment:
I think you should be including an GError as an argument here and returning the reason the parse failed in that rather than setting individual warnings to the daemon log.